component/loki/process/process.go

```diff
@@ -0,0 +1,135 @@
+package process
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/common/loki"
+	"github.com/grafana/agent/component/loki/process/internal/stages"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "loki.process",
+		Args:    Arguments{},
+		Exports: Exports{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments holds values which are used to configure the loki.process
+// component.
+type Arguments struct {
+	ForwardTo []loki.LogsReceiver  `river:"forward_to,attr"`
+	Stages    []stages.StageConfig `river:"stage,block,optional"`
+}
+
+// Exports exposes the receiver that can be used to send log entries to
+// loki.process.
+type Exports struct {
+	Receiver loki.LogsReceiver `river:"receiver,attr"`
+}
+
+var (
+	_ component.Component = (*Component)(nil)
+)
+
+// Component implements the loki.process component.
+type Component struct {
+	opts component.Options
+
+	mut        sync.RWMutex
+	receiver   loki.LogsReceiver
+	fanout     []loki.LogsReceiver
+	processIn  chan<- loki.Entry
+	processOut chan loki.Entry
+	stages     []stages.StageConfig
+}
+
+// New creates a new loki.process component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		opts: o,
+	}
+
+	// Create and immediately export the receiver which remains the same for
+	// the component's lifetime.
+	c.receiver = make(loki.LogsReceiver)
+	o.OnStateChange(Exports{Receiver: c.receiver})
+
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.receiver:
+			c.mut.RLock()
+			select {
+			case <-ctx.Done():
+				return nil
+			case c.processIn <- entry:
+				// no-op
+			}
+			c.mut.RUnlock()
+		case entry := <-c.processOut:
+			c.mut.RLock()
+			for _, f := range c.fanout {
+				select {
+				case <-ctx.Done():
+					return nil
+				case f <- entry:
+					// no-op
+				}
+			}
+			c.mut.RUnlock()
```

Comment:
How many components are going to need to do this? It's a little hard to follow, so if we expect needing to do this more than once, we might want to consider trying to abstract this logic somehow and documenting it in detail to help readers understand what's going on.

Reply:
I don't really expect more components to do so. This is complicated for two reasons: the `processIn` channel has to be send-only (except if we refactor the Loki code) so we couldn't wrap around c.receiver directly, plus each created pipeline running on its own goroutine.

If this pops up again in more logging components (eg. when we bring over the OTel<->Loki adapters) I'd like to see this under a common package eg. similarly to how we do for our own version of Prometheus' Appendable.